Mathlib/RingTheory/LinearDisjoint.lean

```diff
@@ -0,0 +1,603 @@
+  `R`-linearly independent elements `{ a_i }` of `A`, and any family of
+  `R`-linearly independent elements `{ b_j }` of `B`, the family `{ a_i * b_j }` in `S` is
+  also `R`-linearly independent.
+
+- `Subalgebra.LinearDisjoint.of_basis_mul`:
+  conversely, if `{ a_i }` is an `R`-basis of `A`, if `{ b_j }` is an `R`-basis of `B`,
+  such that the family `{ a_i * b_j }` in `S` is `R`-linearly independent,
+  then `A` and `B` are linearly disjoint.
+
+### Other main results
+
+- `Subalgebra.LinearDisjoint.symm_of_commute`, `Subalgebra.linearDisjoint_symm_of_commute`:
+  linearly disjoint is symmetric under some commutative conditions.
+
+- `Subalgebra.LinearDisjoint.bot_left`, `Subalgebra.LinearDisjoint.bot_right`:
+  the image of `R` in `S` is linearly disjoint with any other subalgebras.
+
+- `Subalgebra.LinearDisjoint.sup_free_of_free`: the compositum of two linearly disjoint
+  subalgebras is a free module, if two subalgebras are also free modules.
+
+- `Subalgebra.LinearDisjoint.rank_sup_of_free`,
+  `Subalgebra.LinearDisjoint.finrank_sup_of_free`:
+  if subalgebras `A` and `B` are linearly disjoint and they are
+  free modules, then the rank of `A ⊔ B` is equal to the product of the rank of `A` and `B`.
+
+- `Subalgebra.LinearDisjoint.of_finrank_sup_of_free`:
+  conversely, if `A` and `B` are subalgebras which are free modules of finite rank,
+  such that rank of `A ⊔ B` is equal to the product of the rank of `A` and `B`,
+  then `A` and `B` are linearly disjoint.
+
+- `Subalgebra.LinearDisjoint.adjoin_rank_eq_rank_left`:
+  `Subalgebra.LinearDisjoint.adjoin_rank_eq_rank_right`:
+  if `A` and `B` are linearly disjoint, if `A` is free and `B` is flat (resp. `B` is free and
+  `A` is flat), then `[B[A] : B] = [A : R]` (resp. `[A[B] : A] = [B : R]`).
+  See also `Subalgebra.adjoin_rank_le`.
+
+- `Subalgebra.LinearDisjoint.of_finrank_coprime_of_free`:
+  if the rank of `A` and `B` are coprime, and they satisfy some freeness condition,
+  then `A` and `B` are linearly disjoint.
+
+- `Subalgebra.LinearDisjoint.inf_eq_bot_of_commute`, `Subalgebra.LinearDisjoint.inf_eq_bot`:
+  if `A` and `B` are linearly disjoint, under suitable technical conditions, they are disjoint.
+
+The results with name containing "of_commute" also have corresponding specified versions
+assuming `S` is commutative.
+
+## Tags
+
+linearly disjoint, linearly independent, tensor product
+
+-/
+
+open scoped Classical TensorProduct
+
+open FiniteDimensional
+
+noncomputable section
+
+universe u v w
+
+namespace Subalgebra
+
+variable {R : Type u} {S : Type v}
+
+section Semiring
+
+variable [CommSemiring R] [Semiring S] [Algebra R S]
+
+variable (A B : Subalgebra R S)
+
+/-- If `A` and `B` are subalgebras of `S / R`,
+then `A` and `B` are linearly disjoint, if they are linearly disjoint as submodules of `S`. -/
+@[mk_iff]
+protected structure LinearDisjoint : Prop where
+  submodule : (toSubmodule A).LinearDisjoint (toSubmodule B)
```

Comment:
Any reason why this is a one-field structure and not an abbrev?

I'd suppose a lot of these lemmas will be unnecessary if it is an abbrev.

Reply:
So why not try making it an `abbrev`? Maybe it will reduce a bunch of code duplication? It's hard to get a PR merged if it's very long and relies on definitions which need some discussion.

Reply:
OK let me think and try it again.

Reply:
Seems that it's working; I can't use dot notation in a few places, see commit https://github.com/leanprover-community/mathlib4/pull/15346/commits/0e6961518ae56ef52f77fd8cab40953d6e1cc380